transformer_engine/pytorch/attention.py

```diff
@@ -881,17 +892,55 @@ def get_alibi(
             slopes_shape = torch.Size([1, _alibi_cache["_alibi_slopes"].shape[0], 1, 1])
         if _alibi_cache["_alibi_slopes"].dim() == 2:
             slopes_shape = torch.Size([*_alibi_cache["_alibi_slopes"].shape[:], 1, 1])
-        if bottom_right_alignment:
-            bias = torch.arange(1 - max_seqlen_kv, 1, dtype=torch.int32, device="cuda").view(
-                1, 1, 1, max_seqlen_kv
+        if actual_seqlens_q is None and actual_seqlens_kv is None:
+            if bottom_right_alignment:
+                bias = torch.arange(1 - max_seqlen_kv, 1, dtype=torch.int32, device="cuda").view(
+                    1, 1, 1, max_seqlen_kv
+                )
+            else:
+                bias = torch.arange(
+                    1 - max_seqlen_q,
+                    max_seqlen_kv - max_seqlen_q + 1,
+                    dtype=torch.int32,
+                    device="cuda",
+                ).view(1, 1, 1, max_seqlen_kv)
+            bias = bias - torch.arange(1 - max_seqlen_q, 1, dtype=torch.int32, device="cuda").view(
+                1, 1, max_seqlen_q, 1
             )
+        elif actual_seqlens_q is not None and actual_seqlens_kv is not None:
+            bias = torch.Tensor([]).to(device="cuda")
+            for i, _ in enumerate(actual_seqlens_q):
+                if bottom_right_alignment:
+                    bias_per_batch = torch.arange(
+                        1 - actual_seqlens_kv[i], 1, dtype=torch.int32, device="cuda"
+                    ).view(1, 1, 1, actual_seqlens_kv[i])
+                else:
+                    bias_per_batch = torch.arange(
+                        1 - actual_seqlens_q[i],
+                        actual_seqlens_kv[i] - actual_seqlens_q[i] + 1,
+                        dtype=torch.int32,
+                        device="cuda",
+                    ).view(1, 1, 1, actual_seqlens_kv[i])
+                bias_per_batch = F.pad(
+                    bias_per_batch
+                    - torch.arange(
+                        1 - actual_seqlens_q[i], 1, dtype=torch.int32, device="cuda"
+                    ).view(1, 1, actual_seqlens_q[i], 1),
+                    pad=(
+                        0,
+                        max_seqlen_kv - actual_seqlens_kv[i],
+                        0,
+                        max_seqlen_q - actual_seqlens_q[i],
+                    ),
+                    mode="constant",
+                    value=0,
+                )
+
+                bias = torch.cat([bias, bias_per_batch], dim=0)
```

Comment:
This looks incredibly slow.

Reply:
Hey could you take a look at [0201d6f](https://github.com/NVIDIA/TransformerEngine/pull/1073/commits/0201d6f74645df4854a24d828f0cc155510d860c) please?